### ISSUE TITLE
💅 `noShadow` false positive, flags conflict inside type definition when it shouldn't

### ISSUE THREAD

**Issue by dyc3** [biomejs/biome#6038]
### Environment information

```bash
playground
```

### Rule name

lint/nursery/noShadow

### Playground link

https://next.biomejs.dev/playground/?lintRules=all&analyzerFixMode=safeAndUnsafeFixes&code=dAB5AHAAZQAgAFAAcgBvAHAAcwAgAD0AIAB7AAoAIAAgACAAIABjAG8AbgB0AGEAYwB0ADoAIABDAG8AbgB0AGEAbgB0AEQAYQB0AGEAOwAKACAAIAAgACAAbwBuAEMAaABhAG4AZwBlAD8AOgAgACgAYwBvAG4AdABhAGMAdAA6ACAAQwBvAG4AdABhAG4AdABEAGEAdABhACkAIAA9AD4AIAB2AG8AaQBkADsACgAgACAAIAAgAG8AbgBEAGUAbABlAHQAZQA%2FADoAIAAoAGMAbwBuAHQAYQBjAHQAOgAgAEMAbwBuAHQAYQBuAHQARABhAHQAYQApACAAPQA%2BACAAdgBvAGkAZAA7AAoAfQA7AAoAYwBvAG4AcwB0ACAAewAgAGMAbwBuAHQAYQBjAHQAIAA9ACAAewB9ACwAIABvAG4AQwBoAGEAbgBnAGUAIAA9ACAAKAApACAAPQA%2BACAAewB9ACwAIABvAG4ARABlAGwAZQB0AGUAIAA9ACAAKAApACAAPQA%2BACAAewB9ACAAfQA6ACAAUAByAG8AcABzACAAPQAgAHsAfQA7AAoA&ruleDomains.react=all&ruleDomains.test=all&ruleDomains.solid=all&ruleDomains.next=all

### Expected result

In the playground, `noShadow` flags the function argument in a type definition as shadowing a variable in the upper scope. This shouldn't matter, because the argument names in type definitions don't actually matter for the type definition. 

Since the playground is kinda noisy for this, this is the diagnostic in question.
```
This variable shadows another variable with the same name in the outer scope.
  
    1 │ type Props = {
    2 │     contact: ContantData;
  > 3 │     onChange?: (contact: ContantData) => void;
      │                 ^^^^^^^
    4 │     onDelete?: (contact: ContantData) => void;
    5 │ };
  
  ℹ This is the shadowed variable, which is now inaccessible in the inner scope.
  
    4 │     onDelete?: (contact: ContantData) => void;
    5 │ };
  > 6 │ const { contact = {}, onChange = () => {}, onDelete = () => {} }: Props = {};
      │         ^^^^^^^
    7 │ 
  
  ℹ Consider renaming this variable. It's easy to confuse the origin of variables if they share the same name.
```

See also, eslint playground does not flag this: https://typescript-eslint.io/play#ts=5.7.2&fileType=.tsx&code=C4TwDgpgBACgTgezAZygXigbwFBSgYwQDtgBDfYALimWDgEsiBzAblymIGEALU5iAPzUAFIRLkqNOoyYBKdAD4oANwT0AJmzzEAIhAA2EYIJFiyFarQbN5aJao1sAvmzG0sBYueDooAIj8AGg4iHj4maAxhWyVMKCdg3QMjSKhoxQ8neOp4JFQMTBdsIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uQAsBDACYcA7uiiJo0DtEjgwAXxDygA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by dyc3** [comment]
Sure, all yours.

**Comment by Jordanh1996** [comment]
@dyc3 

I gave this some thought whether the flag is incorrect. Did some GPT consulting and thought about this example:
![Image](https://github.com/user-attachments/assets/69aeb0a6-6382-4317-b150-2cc756180aa1)

No doubt this is weird, and as GPT stated - the Typescript compiler would not shadow anything and use the outer binding. Perhaps there are more practical edge cases I can't think of that might make this a legitimate flagging.

I had some edge case thinking that in some cases within a type definition you might do want to consider shadowing, e.g.:
```ts
type Something = {
  onfunc: (param: { asd: string }) => { onChange: (param: string ) => typeof param };
}
```
The current implementation would flag this. I am unsure about this

WDYT?

**Comment by dyc3** [comment]
See if the eslint version of the rule flags it. We want to try to align with that behavior.

**Comment by arendjr** [comment]
So here's my take: As the rule diagnostic says "It's easy to confuse the origin of variables if they share the same name." This sums up the motivation for the rule pretty well. But for such confusion to take place, there must be a scope in which the shadowed variable can be accessed.

For example:

```ts
const foo = 1;

function bar(foo: number) {
    // This is the scope from where confusion may occur.
    // If someone looks hastily and sees the global variable,
    // but overlooks the function argument, they may get
    // the wrong impression.
}
```

Now back to the issue at hand:

```ts
type Props = {
  contact: string;
  onChange?: (contact: string) => void;
  onDelete?: (contact: string) => void;
};
const { contact = "", onChange = () => { }, onDelete = () => { } }: Props = {};
```

`onChange` and `onDelete` don't have a function body, and so there _is_ no scope in which this confusion can take place. Therefore I agree, these should be considered false positives.
